elliptic-curve/src/group_digest.rs

```diff
@@ -0,0 +1,86 @@
+use crate::hash2field::{hash_to_field, ExpandMsg, FromOkm};
+use crate::isogeny::Isogeny;
+use crate::osswu::OsswuMap;
+use ff::Field;
+use group::Group;
+
+/// Convert a set of field elements into a point
+pub trait ConvertToPoint {
+    /// The field element that represents a point
+    type FieldElement: Field;
+    /// The result of converting two field elements to a Point
+    type Output;
+
+    /// Converts two coordinates into a point
+    fn convert_to_point(x: Self::FieldElement, y: Self::FieldElement) -> Self::Output;
+}
+
+/// Method for points to be cleared by the curve cofactor
+pub trait ClearCofactor {
+    /// Multiply this point by the curve cofactor to ensure
+    /// correct subgroup membership
+    fn clear_cofactor(&mut self);
+}
+
+/// Adds hashing arbitrary byte sequences to a valid group element
+pub trait GroupDigest {
+    /// The field element representation for a group value with multiple elements
+    type FieldElement: FromOkm + OsswuMap + Isogeny;
+    /// The resulting group element
+    type Output: Group
+        + ClearCofactor
+        + ConvertToPoint<FieldElement = Self::FieldElement, Output = Self::Output>;
+
+    /// Computes the hash to curve routine according to
+    /// <https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html>
+    /// which says
+    /// Uniform encoding from byte strings to points in G.
+    /// That is, the distribution of its output is statistically close
+    /// to uniform in G.
+    /// This function is suitable for most applications requiring a random
+    /// oracle returning points in G assuming a cryptographically secure
+    /// hash function is used.
+    ///
+    /// Example using a fixed size hash function
+    ///
+    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXmd<sha2::Sha256>>(b"test data", b"CURVE_XMD:SHA-256_SSWU_RO_");
+    ///
+    ///
+    /// Example using an extendable output function
+    ///
+    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXof<sha3::Shake256>>(b"test data", b"CURVE_XOF:SHAKE-256_SSWU_RO_");
```

Comment:
```suggestion
    /// # Examples
    /// Using a fixed size hash function:
    /// ```
    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXmd<sha2::Sha256>>(b"test data", b"CURVE_XMD:SHA-256_SSWU_RO_");
    /// ```
    ///
    /// Using an extendable output function:
    /// ```
    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXof<sha3::Shake256>>(b"test data", b"CURVE_XOF:SHAKE-256_SSWU_RO_");
    /// ```
```